src/main/java/net/minecraftforge/common/ForgeSpawnEggItem.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common;
+
+import net.minecraft.core.Direction;
+import net.minecraft.core.dispenser.DispenseItemBehavior;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.MobSpawnType;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.SpawnEggItem;
+import net.minecraft.world.level.block.DispenserBlock;
+import net.minecraft.world.level.gameevent.GameEvent;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class ForgeSpawnEggItem extends SpawnEggItem
+{
+    private static final List<ForgeSpawnEggItem> MOD_EGGS = new ArrayList<>();
+    private static final Map<EntityType<? extends Mob>, ForgeSpawnEggItem> TYPE_MAP = new IdentityHashMap<>();
+    private final Supplier<? extends EntityType<? extends Mob>> typeSupplier;
+
+    public ForgeSpawnEggItem(Supplier<? extends EntityType<? extends Mob>> type, int backgroundColor, int highlightColor, Properties props)
+    {
+        super((EntityType<? extends Mob>) null, backgroundColor, highlightColor, props);
+        this.typeSupplier = type;
+
+        MOD_EGGS.add(this);
+    }
+
+    @Override
+    public EntityType<?> getType(@Nullable CompoundTag tag)
+    {
+        EntityType<?> type = super.getType(tag);
+        return type != null ? type : typeSupplier.get();
+    }
+
+    @Nullable
+    protected DispenseItemBehavior createDispenseBehavior()
+    {
+        return DEFAULT_DISPENSE_BEHAVIOR;
+    }
+
+    @Nullable
+    public static ForgeSpawnEggItem fromEntityType(@Nullable EntityType<?> type)
+    {
+        return TYPE_MAP.get(type);
+    }
+
+    public static Iterable<ForgeSpawnEggItem> getModEggs()
+    {
+        return MOD_EGGS;
+    }
+
+    public static void finalizeModEggs()
```

Comment:
This method should either be commented as 'For internal use only'
Or moved to a private method that is auto resisted using @EventBusSubscriber.
So that it doesn't exist from a API standpoint.